account_payment_purchase/tests/test_account_payment_purchase.py

```diff
@@ -84,8 +84,8 @@ def test_purchase_order_invoicing(self):
         invoice = self.env["account.move"].create(
             {"partner_id": self.partner.id, "move_type": "in_invoice"}
         )
-        with Form(invoice) as inv:
-            inv.purchase_id = self.purchase
+        invoice.purchase_id = self.purchase
```

Comment:
Why are you removing the form? It was done this way in order to keep the same logic done by odoo interface

Reply:
@etobella AssertionError: can't write on invisible field purchase_id

Reply:
But with this, you are not testing the real thing, as onchanges are not triggered.

Reply:
Having a quick look at how is implemented in core: it sucks. It seems to go against the principle of not having business logic in onchanges. Unfortunately, if you don't set `force_save="1"` on the field it does not work and - BTW - the field is emptied at the end of the onchange.
Also, if you look at `_set_purchase_orders` they call this onchange explicitly :facepalm: 

However, if you want to keep it like this you should call `_onchange_partner_id`  too IMO.

Reply:
@simahawk

If I'm not mistaken, the function that is called when the purchase_id field is modified is _onchange_purchase_auto_complete

https://github.com/odoo/odoo/blob/08cf96d9ba169287a7e8d7dece4ac0110fa01742/addons/purchase/models/account_invoice.py#L33:L34

And this is already being called in the test, exactly the same as in the rest of the tests that I have not modified.

https://github.com/OCA/bank-payment/blob/1102ba6ed053e3cbe708512d8e71d0535aad0bc0/account_payment_purchase/tests/test_account_payment_purchase.py#L88

Isn't that correct?